plugins/catalog-backend/src/ingestion/processors/GitlabReaderProcessor.test.ts

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { getVoidLogger } from '@backstage/backend-common';
+import { LocationSpec } from '@backstage/catalog-model';
+import { ConfigReader } from '@backstage/config';
+import {
+  getApiRequestOptions,
+  getApiUrl,
+  getRawRequestOptions,
+  getRawUrl,
+  GitLabReaderProcessor,
+  ProviderConfig,
+  readConfig,
+} from './GitlabReaderProcessor';
+
+describe('GithubReaderProcessor', () => {
+  describe('getApiRequestOptions', () => {
+    it('sets the correct API version', () => {
+      const config: ProviderConfig = { target: '', apiBaseUrl: '' };
+      expect((getApiRequestOptions(config).headers as any).Accept).toEqual(
+        'application/json',
+      );
+    });
+
+    it('inserts a token when needed', () => {
+      const withToken: ProviderConfig = {
+        target: '',
+        apiBaseUrl: '',
+        token: 'A',
+      };
+      const withoutToken: ProviderConfig = {
+        target: '',
+        apiBaseUrl: '',
+      };
+      expect(
+        (getApiRequestOptions(withToken).headers as any)['PRIVATE-TOKEN'],
+      ).toEqual('A');
+      expect(
+        (getApiRequestOptions(withoutToken).headers as any)['PRIVATE-TOKEN'],
+      ).toBeUndefined();
+    });
+  });
+
+  describe('getRawRequestOptions', () => {
+    it('inserts a token when needed', () => {
+      const withToken: ProviderConfig = {
+        target: '',
+        rawBaseUrl: '',
+        token: 'A',
+      };
+      const withoutToken: ProviderConfig = {
+        target: '',
+        rawBaseUrl: '',
+      };
+      expect(
+        (getRawRequestOptions(withToken).headers as any)['PRIVATE-TOKEN'],
+      ).toEqual('A');
+      expect(
+        (getRawRequestOptions(withoutToken).headers as any)['PRIVATE-TOKEN'],
+      ).toBeUndefined();
+    });
+  });
+
+  describe('getApiUrl', () => {
+    it('rejects targets that do not look like URLs', () => {
+      const config: ProviderConfig = { target: '', apiBaseUrl: '' };
+      expect(() => getApiUrl('a/b', config)).toThrow(/Incorrect URL: a\/b/);
+    });
+
+    it('happy path for gitlab', () => {
+      const config: ProviderConfig = {
+        target: 'https://gitlab.com',
+        apiBaseUrl: 'https://gitlab.com/api/v4',
+      };
+      expect(
+        getApiUrl(
+          'https://gitlab.com/a/b/blob/branchname/path/to/c.yaml',
```

Comment:
Hey the original tests had some explicit tests for a URL format that has an extra `/-/` in there - apparently it's because of an old style and a new stile URL pattern after they upgraded Gitlab. Could you make sure to copy over that pattern and test BOTH of them and make sure they both output the same v4 URL? We have a user that's specifically struggling with this duality at the moment. Thanks

Reply:
This new code does work with the old format including `/-/` because the parse-git-url library supports it. I think it may be for repos that are under groups/subgroups - [example](https://gitlab.com/forrest-backstage/backstage-demo/-/blob/master/hello-world.yaml). Agree it would be nice to have tests for both :)